### PR TITLE
BUG: support dtypes in column_dtypes for to_records()

### PR DESCRIPTION
**Pull request by qwhelan**
Despite the name, the `column_dtypes` param in `DataFrame.to_records()` doesn't accept `np.dtype` objects:
```
In [3]: df = pd.DataFrame([[0, 1], [2, 3]])
In [4]: df.to_records(column_dtypes=df.dtypes)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-e3541f39084e> in <module>()
----> 1 df.to_records(column_dtypes=df.dtypes)

~/projects/pandas/pandas/core/frame.py in to_records(self, index, convert_datetime64, column_dtypes, index_dtypes)
   1724                 msg = ("Invalid dtype {dtype} specified for "
   1725                        "{element} {name}").format(dtype=dtype_mapping,
-> 1726                                                   element=element, name=name)
   1727                 raise ValueError(msg)
   1728 

ValueError: Invalid dtype int64 specified for column 0
```
The above also fails when `df.dtypes` is cast into a `dict` for overriding.

This PR simply adds `np.dtype` to the accepted types and adopts a subset of the tests that currently cover the numpy scalar types (`np.int64` and similar, which are not dtypes).

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
cc @jreback 

@qwhelan : Thanks for the PR!  I do see however that our docstring explicitly does not mention `np.dtype` as an accepted data type for this parameter (see [here](https://github.com/pandas-dev/pandas/blob/a86d008be325fea200a9c5f328e77f652d69045e/pandas/core/frame.py#L1566)).  This is thus not a bug, but an enhancement that you are proposing.

**Comment by qwhelan** [comment]
@gfyoung My initial thought was that this was an enhancement until I read the discussion on #22229, where the intent seemed to be to support `np.dtype` as well: https://github.com/pandas-dev/pandas/pull/22229/files#r244338742
> type should be `numpy.dtype` right? Oh, but we probably also accept things like `int`, which is also a type.
> Yep, that’s why I use type and not numpy.dtype.

Entirely possible I'm reading it wrong, but it looks like there was an assumption that `type` would handle `np.dtype` and `int`/etc (it only covers the latter). Additional evidence here is that the parameter name includes `dtype`, the docstring references `data types`, and the function itself uses `np.dtype` as a valid input:
```python
if dtype_mapping is None:
    formats.append(v.dtype)
elif isinstance(dtype_mapping, (type, compat.string_types)):
    formats.append(dtype_mapping)
...
```

**Comment by gfyoung** [comment]
xref: https://github.com/pandas-dev/pandas/pull/22229#discussion_r244338742

Fair enough.  I'm not opposed to adding this *per se*.  It was more of a question of semantics regarding bug and enhancement.

File: pandas/tests/frame/test_convert_to.py [hunk 1]
```diff
@@ -274,6 +274,29 @@ def test_to_records_dtype(self, kwargs, expected):
...
+        # See gh-24895
+        df = DataFrame({"A": [1, 2], "B": [0.2, 1.5], "C": ["a", "bc"]})
+
+        if isinstance(expected, str):
+            with pytest.raises(ValueError, match=expected):
+                df.to_records(**kwargs)
+        else:
+            result = df.to_records(**kwargs)
+            tm.assert_almost_equal(result, expected)
+
```
Comment (gfyoung) [inline pandas/tests/frame/test_convert_to.py]: I think you can just add another parameter to the tests that I already have.

Comment (qwhelan) [reply pandas/tests/frame/test_convert_to.py]: Sure, will do.

File: pandas/tests/frame/test_convert_to.py [hunk 2]
```diff
@@ -274,6 +274,29 @@ def test_to_records_dtype(self, kwargs, expected):
             result = df.to_records(**kwargs)
             tm.assert_almost_equal(result, expected)
 
+    @pytest.mark.parametrize("kwargs,expected", [
```
Comment (jreback) [inline pandas/tests/frame/test_convert_to.py]: do we test with EA types / invalid types (misspelled for example)? These should just raise as numpy can't support them. (it probably does, but do we have a test for it)?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_convert_to.py (hunk 2 above)
Comment (qwhelan) [reply pandas/tests/frame/test_convert_to.py]: Looks like both cases do actually raise, and I've added cases to cover them (former is a `ValueError`, latter is a `TypeError`)

File: pandas/tests/frame/test_convert_to.py (hunk 1 above)
Comment (qwhelan) [reply pandas/tests/frame/test_convert_to.py]: Done

**Comment by jreback** [comment]
lgtm. can you add a whatsnew note (0.25.0), bug fixes reshaping. @gfyoung if you can have a look.

**Comment by qwhelan** [comment]
@jreback Done

**Review by gfyoung — APPROVED** [review]
Nice!

**Comment by gfyoung** [comment]
@jreback : This looks good to me.  Not sure if you wanted to have a look once more?

**Review by jreback — APPROVED** [review]

**Comment by qwhelan** [comment]
Thanks! Small perf improvement built on this coming soon

**Pull request merged**